### PR TITLE
Issue154

### PR DESCRIPTION
**Pull request by ahmedqarmout2**
I dont know what this one was about. But I fixed a lot of things

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -563,16 +573,49 @@ app.post('/submitanswer', function(req, res) {
...
-        function(err, value) {
-            var result = value.correct ? value : 'incorrect';
-            var status = value.correct ? 200 : 500;
-            return res.status(status).send(result);
+    var questionId = req.body.questionId;
+    var answer = req.body.answer;
+    var userId = req.session.user.id;
+
+    questions.lookupQuestionById(questionId,function(err, question){
+		if(err){
```
Comment (esteeroramy) [inline server.js]: space

File: server/questions.js [hunk 2]
```diff
@@ -159,79 +159,99 @@ exports.lookupQuestionById = function(questionId, callback) {
...
-	var userId = user.id;
-
-	lookupQuestionById(questionId, function(err, question){
-		if(err || !question){
-			return callback('error', null);
-		}
+// adding rating to question collection
+exports.submitRating = function (questionId, userId, rating, callback) {
+	var currentDate = new Date().toString();
+    var query = {_id: questionId};
```
Comment (esteeroramy) [inline server/questions.js]: spaces

File: server/questions.js [hunk 3]
```diff
@@ -159,79 +159,99 @@ exports.lookupQuestionById = function(questionId, callback) {
...
+}
+
+// get all questions list
+exports.getAllQuestionsList = function(callback) {
+	db.getQuestionsList({}, {id: 1}, callback);
+}
+
+// submit answer
+exports.submitAnswer = function(questionId, userId, correct, points, answer, callback) {
+	var currentDate = new Date().toString();
```
Comment (esteeroramy) [inline server/questions.js]: spaces

File: server/questions.js [hunk 4]
```diff
@@ -159,79 +159,99 @@ exports.lookupQuestionById = function(questionId, callback) {
...
+    var query = {_id: questionId};
+    var update = {};
+
+    update.$push = {};
+    update.$inc = {};
+
+	query['correctAttempts.id'] = { $ne : userId };
```
Comment (esteeroramy) [inline server/questions.js]: spaces

**Review by esteeroramy — APPROVED** [review]

File: dataGenerator.js [hunk 5]
```diff
@@ -140,6 +140,40 @@ var answerQuestion = function(questionId) {
     }
 }
 
+// check answer
+var checkAnswer = function (questionId, userId, answer, callback) {
+    logger.info('User %s attempted to answer question %s with "%s"', userId, questionId, answer);
+    db.lookupQuestion({id: questionId}, function(err, question) {
+        questionId = question._id;
+        if(err){
```
Comment (Mohammed-JS-Khan) [inline dataGenerator.js]: Space before open curly

File: public/js/admin.js [hunk 6]
```diff
@@ -581,7 +581,7 @@ var deleteQuestion = function(qid) {
 
 var editQuestion = function(qid) {
     $.ajax({
-        type: 'POST',
+        type: 'GET',
```
Comment (Mohammed-JS-Khan) [inline public/js/admin.js:584]: good call. the function name is missleading lol. GET is the correct type though.

File: server.js [hunk 7]
```diff
@@ -563,16 +573,49 @@ app.post('/submitanswer', function(req, res) {
...
+        var points = question.points;
+        var text = value ? 'correct' : 'incorrect';
+        var status = value ? 200 : 500;
+        var response = {text: text, points: points};
+
+        if (req.session.user.type === common.userTypes.ADMIN) {
+            return res.status(status).send(response);
+        }
+
+        users.submitAnswer(userId, questionId, value, points, answer, function(err, result){
```
Comment (Mohammed-JS-Khan) [inline server.js:603]: Space before open curly

File: server.js [hunk 8]
```diff
@@ -563,16 +573,49 @@ app.post('/submitanswer', function(req, res) {
...
+            if(err){
```
Comment (Mohammed-JS-Khan) [inline server.js:604]: Space before open curly

File: server.js [hunk 9]
```diff
@@ -563,16 +573,49 @@ app.post('/submitanswer', function(req, res) {
...
+                logger.error(err);
+                return res.status(500).send(err);
+            }
+
+            questions.submitAnswer(questionId, userId, value, points, answer, function(err, result) {
+                if(err){
```
Comment (Mohammed-JS-Khan) [inline server.js:610]: Space before open curly

File: server.js (hunk 7 above)
Comment (ahmedqarmout2) [reply server.js:603]: I will kill you Mo. Just approve

File: server/questionValidator.js [hunk 10]
```diff
@@ -24,133 +24,133 @@ const successMsg = {success:true, msg:'Validation Passed'}
...
+            extraAttributes = true;
+        }
+    }
+
+    // check by question type to validate the extra fields
+    if (extraAttributes){
+        return validateQuestionAttributesByType(question,type);
+    }
+    return successMsg;
 }
```
Comment (Mohammed-JS-Khan) [inline server/questionValidator.js:61]: idk if you wanna fix the spaces before open curly

**Review by Mohammed-JS-Khan — APPROVED** [review]
Lots of missing spaces before open curly brackets but i understand this is urgent so lets move this along. Nice revamp.

**Pull request merged**